source/extensions/filters/common/expr/evaluator.cc

```diff
@@ -138,14 +143,66 @@ BuilderInstanceSharedPtr createBuilder(Protobuf::Arena* arena) {
     throw CelException(absl::StrCat("failed to register extension regex functions: ",
                                     ext_register_status.message()));
   }
-  return std::make_shared<BuilderInstance>(std::move(builder));
+  // Register string extension functions only if enabled in configuration.
+  if (enable_string_functions) {
+    auto string_register_status =
+        cel::extensions::RegisterStringsFunctions(builder->GetRegistry(), options);
+    if (!string_register_status.ok()) {
+      throw CelException(absl::StrCat("failed to register extension string functions: ",
+                                      string_register_status.message()));
+    }
+  }
+  return builder;
 }
 
-SINGLETON_MANAGER_REGISTRATION(expression_builder);
+// Cache to store builders for different configurations.
+class BuilderCache : public Singleton::Instance {
+public:
+  using ConfigHash = size_t;
+
+  BuilderInstanceSharedConstPtr
+  getOrCreateBuilder(OptRef<const envoy::config::core::v3::CelExpressionConfig> config) {
+    ASSERT_IS_MAIN_OR_TEST_THREAD();
+
+    ConfigHash hash = 0;
+    if (config.has_value()) {
+      // Use MessageUtil::hash for proto hashing.
+      hash = MessageUtil::hash(config.ref());
+    }
+
+    auto it = builders_.find(hash);
+    if (it != builders_.end()) {
+      return it->second;
+    }
+
+    // Create new builder with the configuration.
+    auto builder = createBuilder(config);
+    auto instance = std::make_shared<BuilderInstance>(std::move(builder));
+    builders_[hash] = instance;
+    return instance;
+  }
+
+private:
+  absl::flat_hash_map<ConfigHash, BuilderInstanceSharedConstPtr> builders_;
```

Comment:
Shouldn't this be a weak_ptr like it was with singleton? I think it needs to be able to be released after xDS is unloaded.